base/parse.jl

```diff
@@ -227,14 +227,16 @@ end
 end
 
 """
-    tryparse(type, str; base)
+    tryparse(type, str, default; base)
 
 Like [`parse`](@ref), but returns either a value of the requested type,
 or [`nothing`](@ref) if the string does not contain a valid number.
+Optionally you can specify a fallback value to return instead of [`nothing`](@ref).
 """
-function tryparse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = nothing) where {T<:Integer}
+function tryparse(::Type{T}, s::AbstractString, default = nothing; base::Union{Nothing,Integer} = nothing) where {T<:Integer}
     # Zero base means, "figure it out"
-    tryparse_internal(T, s, firstindex(s), lastindex(s), base===nothing ? 0 : check_valid_base(base), false)
+    val = tryparse_internal(T, s, firstindex(s), lastindex(s), base===nothing ? 0 : check_valid_base(base), false)
+    isnothing(default) ? val : something(val, default)
```

Comment:
AFAICT this should be either `something(val, default)` or `default === nothing ? val : default`, not both. The latter is safer if `T` can be `<: Some`. (`=== nothing` is better than `isnothing` for inference)

Have you considered the other approach, which is to pass `default` to `tryparse_internal` and have it return that instead of `nothing`? It could be slightly faster (not sure).

Reply:
`something(val, default)` errors if both `val` and `default` are `nothing`, and `default === nothing ? val : default` will always return `default` if one is passed in.

I was hoping to avoid clutter by not propagating to `tryparse_internal`, but I'll run some benchmarks to see if there is a speedup.

Reply:
> `something(val, default)` errors if both `val` and `default` are `nothing`, and `default === nothing ? val : default` will always return `default` if one is passed in.

Ah, yes, of course. Though using `default === nothing` instead of `isnothing(default)` might still be better for the compiler.

> I was hoping to avoid clutter by not propagating to `tryparse_internal`, but I'll run some benchmarks to see if there is a speedup.

I can't be sure without trying, but I had the impression that adding the argument to `tryparse` could actually be cleaner, as it would just require replacing `nothing` with `default` everywhere, instead of adding special code to replace `nothing` after the fact. But I might be wrong.

Reply:
I've added a commit with these changes but didn't observe any significant difference once compiled into a sysimage.

Another thing to consider is whether we should enforce `default::Union{T,Missing,Nothing}` or allow `default::Any`, in which case we would need a workaround for the `Float16` implementation.

Reply:
> I've added a commit with these changes but didn't observe any significant difference once compiled into a sysimage.

Yeah, I wasn't saying it would necessarily make a difference in this case, it's just better in general.

> Also, tests are now failing as some stdlibs depend on `Base.tryparse_internal`, so I'm not convinced adding the argument to `tryparse_internal` is a good idea.

Why not use `default = nothing` in `tryparse_internal` so that it keeps returning `nothing` if you don't provide the argument?

> Another thing to consider is whether we should enforce `default::Union{T,Missing,Nothing}` or allow `default::Any`, in which case we would need a workaround for the `Float16` implementation.

What's the problem with `Float16`?

Reply:
>Why not use `default = nothing` in `tryparse_internal` so that it keeps returning `nothing` if you don't provide the argument?

Then you have a default value for `default` in both `tryparse` and `tryparse_internal`, which means you could accidentally stop passing `default` from `tryparse` to `tryparse_internal` in the future and not see any issues. Although I suppose that's mitigated by having a broad range of tests.

>What's the problem with `Float16`?

```
tryparse(::Type{Float16}, s::AbstractString) =
    convert(Union{Float16, Nothing}, tryparse(Float32, s))
```

`Float16` is "parsed" by parsing as a `Float32` and then `convert`ing to `Float16`, so this would need something like `val === nothing ? default : val`. But from the above conversation

>We know that the value should be a Float64, but we unfortunately don't know what the value is since the input was bad.

I think we might want to annotate the `default`s with `::Union{T,Missing,Nothing}` to be type-stable.

Reply:
> Then you have a default value for `default` in both `tryparse` and `tryparse_internal`, which means you could accidentally stop passing `default` from `tryparse` to `tryparse_internal` in the future and not see any issues. Although I suppose that's mitigated by having a broad range of tests.

Right, but I don't think that's a problem. As you say, tests are there to check that it works.

> `Float16` is "parsed" by parsing as a `Float32` and then `convert`ing to `Float16`, so this would need something like `val === nothing ? default : val`.

Yeah, but I guess it's OK as long as it's only in one method.

> I think we might want to annotate the `default`s with `::Union{T,Missing,Nothing}` to be type-stable.

Why would this affect affect type stability? :-/

Reply:
> Why would this affect affect type stability? :-/

I guess `tryparse` isn't a type-stable function to begin with since it could return `nothing` (maybe I'm misunderstanding type stability?), but if `tryparse` returns a value that is neither `nothing` or `missing`, shouldn't it be of type `T`? Or can we assume the user supplying `default` will handle it?

Reply:
`tryparse(T, str, default)` should be inferred by the compiler as returning `Union{T, typeof(default)}`, and yes, that's the user's job to handle the `default` he provided.